crypto/crypto.go

```diff
@@ -37,8 +37,23 @@ func RandomBytes(s []byte) {
 	}
 }
 
+// GenerateAddressFromSK take a secret key and returns the corresponding Address
+func GenerateAddressFromSK(sk []byte) (types.Address, error) {
+	edsk := ed25519.PrivateKey(sk)
+
+	var a types.Address
+	pk := edsk.Public()
+	n := copy(a[:], []byte(pk.(ed25519.PublicKey)))
+	if n != ed25519.PublicKeySize {
+		return [32]byte{}, fmt.Errorf("generated public key has the wrong size, expected %d, got %d", ed25519.PublicKeySize, n)
+	}
+	return a, nil
+}
+
 // SignTransaction accepts a private key and a transaction, and returns the
 // bytes of a signed transaction ready to be broadcasted to the network
+// If the SK's corresponding address is different than the txn sender's, the SK's
+// corresponding address will be assigned as AuthAddr
```

Comment:
This is really elegant, I really that you were able to implement this without changing the method signature.

I can also imagine someone grabbing the wrong private key and being annoyed that the transaction failed because the account wasn't actually rekeyed. If there was a separate `SignTransactionAuthAddr`, then this one could return an error if you tried to sign the tx with a key other than the sender.

I'm not sure which approach is best, this is probably fine.

Reply:
You're right and I also debated that when adding it. The reason I chose this way is (1) to keep it consistent with Adam's implementation and (2) I believe that the majority of cases will be a success so it's better to make it more convenient for the majority rather than the opposite.